prov/shm/src/smr_ep.c

```diff
@@ -132,12 +132,28 @@ int smr_setopt(fid_t fid, int level, int optname,
 	struct smr_ep *smr_ep =
 		container_of(fid, struct smr_ep, util_ep.ep_fid);
 
-	if ((level != FI_OPT_ENDPOINT) || (optname != FI_OPT_MIN_MULTI_RECV))
+	if ((level != FI_OPT_ENDPOINT))
 		return -FI_ENOPROTOOPT;
 
-	smr_get_smr_srx(smr_ep)->min_multi_recv_size = *(size_t *)optval;
+	if (optname == FI_OPT_MIN_MULTI_RECV) {
+		smr_get_smr_srx(smr_ep)->min_multi_recv_size = *(size_t *)optval;
+		return FI_SUCCESS;
+	}
 
-	return FI_SUCCESS;
+	if (optname == FI_OPT_CUDA_API_PERMITTED) {
+		if (!hmem_ops[FI_HMEM_CUDA].initialized) {
+			FI_WARN(&smr_prov, FI_LOG_CORE,
+				"Cannot set option FI_OPT_CUDA_API_PERMITTED when cuda library "
+				"or cuda device not available\n");
+			return -FI_EINVAL;
+		}
+
+		/* our CUDA support relies on the ability to call CUDA API */
```

Comment:
@wzamazon Clarifying question about this comment - shm relies on the CUDA API for the IPC path or something else? Can we disable p2p and use gdrcopy instead?

Reply:
I think it's only required by IPC path. But currently shm will IPC for hmem as long as it is available. Shm can use gdrcopy for non-IPC protocols. But we have not implemented it yet.

Reply:
To be clear, to use gdrcopy, SHM provider need to register memory with gdrcopy. which it does not do yet.

Reply:
That makes sense, thank you!